### ISSUE TITLE
DevEx: Calling "`readSignal`" with arguments should throw in solid-js dev mode

### ISSUE THREAD

**Issue by danieltroger** [solidjs/solid#2045]
### Describe the bug

Right now it's possible to do a mistake when destructuring, see this example: https://playground.solidjs.com/anonymous/26548b4b-87ed-4b76-bfbd-2cb6fa2b5b42

This mistake is not caught by TypeScript right now and leads to unexpected behavior - effects being re-executed when nothing happens and my value not being set.

### Your Example Website or App

https://playground.solidjs.com/anonymous/26548b4b-87ed-4b76-bfbd-2cb6fa2b5b42

### Steps to Reproduce the Bug or Issue

1. Go to playground link
2. See that a getter gets passed to a setter without any error

### Expected behavior

I as a dev should not have to debug such a basic mistake, my computer should blurt red in my face.
Since it seems like nothing happens with stuff passed to a `readSignal` function right now it would be easy for you to just throw in dev mode, saving people that make this mistake time debugging.

### Platform

macOS (M1), latest chrome or firefox

**Comment by ryansolid** [comment]
Is this a TS request/error? It seems like it. We use setters with arguments in a number of places internally I believe but that is only for undefinable values. Like if you initialize with a number and the Signal is not able to be set to `undefined` I agree that there should be an issue here.

**Comment by fabiospampinato** [comment]
I think the request is for throwing at runtime when calling a getter with more than 0 arguments, so nothing really strictly about types 🤔

**Comment by danieltroger** [comment]
> I think the request is for throwing at runtime when calling a getter with more than 0 arguments, so nothing really strictly about typescript types 🤔

Exactly this

**Comment by ryansolid** [comment]
Ok.. that seems reasonable. Obviously, dev mode is no cost.  Prod is a different beast.

This issue no-ops right now so I wouldn't be opposed to warning in dev rather than erroring. Erroring is stronger and suggests we should break, but then again I don't know that I'd want to put these sort of error checks in many places in general. Like it isn't really the libraries responsibility to check for people calling functions with extra arguments. It would be different validating inputs but its like outside of the scope of operation. I see value in aiding in Dev but it seems oddly forced for prod. But throw worthy error feels it should always throw.

This is why I went to TS as that's what its for. The TS issue which I assume is because argumentless setters which admittedly might be a stupid thing. I cared a lot about the characters so people working TS had to work around that. It is just something in that domain.

**Comment by danieltroger** [comment]
> I cared a lot about the characters so people working TS had to work around that.

You mean the bundle size or the subjective character of signals?

> Like it isn't really the libraries responsibility to check for people calling functions with extra arguments.

I think that if solid-js wants to gain more adoption it should keep devEx in mind and "close" common traps one may fall into. It's a library many humans will use, it should account for them making mistakes. While this cost me 10 minutes, it may cause someone starting to learn solid to just give up on solid-js entirely, thinking it's "too complicated".

But I totally buy that the throw is not worth the bundle size in prod. Just add a warning in dev mode if you think that's the best and I'm happy too.

**Comment by ryansolid** [comment]
Yeah bundle size.. setting `set(undefined)` becomes `s(void 0)` in most modification rather than just `s()`. I don't care as much now. But historically it was a thing. I'd love to simplify Setter types if it would help in the future. Although I think the TS issue is more complicated than that when reading the conversation.

Yeah let's go with dev warning.

**Comment by ryansolid** [comment]
I attempted to add this to the repo and it actually caused a bunch of internal bugs. Not too hard to fix. But requires adding function wrappers on `createMemo` calls mostly. It is interesting because unlike with the write there is no really downside to reading a Accessor read into a `createMemo` call directly. Which leaves me a bit on the fence here on erroring in this case.

**Comment by danieltroger** [comment]
Did you attempt [the typescript fix](https://www.typescriptlang.org/play/?#code/JYWwDg9gTgLgBAbzgYygUwIYzQZWAcwDsMAbOAXzgDMoIQ4ByAZwhOABMBaAKyYYG4AUIJgBPMGjg40MbFABqpAK5oA8lADCAIwA8AFQB8cALxw9cAD5wAFNbDoAbgC4zAShNG9roWIlSZcvpGptZKYOxYaFAu0rJRiiQq6tpB7sZGDhAcPuKSAILIyGhMLFBBJjYAdNUYUPhMLoRoDlFpnkKCyBCETPAA2vgBUQC6FaiY2HhEpDqESiBaUQbWrp3dvXDYvS4FRSXQs-OLUME2bXAAzB1USoTIMMDdmxgA1sV5TENQ1p9x0f5-Q4LJbuBDCcjCGCvd6-OTWQZ-VZQt5MD5faxbGCuIA) ([discord link](https://discord.com/channels/722131463138705510/751355413701591120/1199397127143555233)) or the dev warning?

I assume the typescript fix and it makes so that it's illegal to call accessors with arguments? 🤔

**Comment by ryansolid** [comment]
Both. In a sense the TS highlights the issue that we would run into.

This is a real footgun, but throwing from accessors when they receive arguments would break an intentional part of Solid’s function ergonomics. Especially with 2.0 I think.

Accessors are zero-argument by contract, but JavaScript functions commonly ignore extra arguments, and Solid APIs rely on that being safe. For example, passing an accessor directly as the compute/front half of an effect or memo-like primitive can be valid even if the caller supplies a previous value argument internally.

So a runtime `arguments.length` check would catch this destructuring mistake, but it would also reject valid callback usage. That makes it too broad for core.

I think this is better handled by tooling, documentation, or a lint rule that can detect likely accessor/setter mixups in context without changing accessor call semantics.
